src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs

```diff
@@ -48,20 +62,40 @@ public PdbSourceDocumentLoaderService()
             return Task.FromResult<TextLoader?>(null);
         }
 
-        private static TextLoader? TryLoadSourceFromDisk(SourceDocument sourceDocument)
+        private static Stream? TryGetEmbeddedSourceStream(SourceDocument sourceDocument)
         {
-            using var fileStream = new FileStream(sourceDocument.FilePath, FileMode.Open, FileAccess.Read, FileShare.Read | FileShare.Delete);
-            if (fileStream is null)
+            if (sourceDocument.EmbeddedTextBytes is null)
                 return null;
 
-            // TODO: Don't hard code UTF8: https://github.com/dotnet/roslyn/issues/57350
-            var sourceText = SourceText.From(fileStream, Encoding.UTF8, sourceDocument.HashAlgorithm);
-            var fileChecksum = sourceText.GetChecksum();
+            var embeddedTextBytes = sourceDocument.EmbeddedTextBytes;
+            var uncompressedSize = BitConverter.ToInt32(embeddedTextBytes, 0);
+            var stream = new MemoryStream(embeddedTextBytes, sizeof(int), embeddedTextBytes.Length - sizeof(int));
+
+            if (uncompressedSize != 0)
+            {
+                var decompressed = new MemoryStream(uncompressedSize);
+
+                using (var deflater = new DeflateStream(stream, CompressionMode.Decompress))
+                {
+                    deflater.CopyTo(decompressed);
+                }
 
-            if (fileChecksum.SequenceEqual(sourceDocument.Checksum))
+                if (decompressed.Length != uncompressedSize)
+                {
+                    return null;
+                }
+
+                stream = decompressed;
+            }
+
+            return stream;
+        }
+
+        private static Stream? TryGetFileStream(SourceDocument sourceDocument)
+        {
+            if (File.Exists(sourceDocument.FilePath))
```

Comment:
Is File.Exists guaranteed to not throw, even if the path is corrupted or something?

Reply:
The documentation doesn't say the word "guarantee", but thats certainly how it reads to me :)